### PR TITLE
Skip unsupported type widening columns

### PR DESCRIPTION
**Pull request by SemionPar**
## Description
A change from failing to skipping is for flexibility, because Databricks supports wider variety of type changes.

Draft todos:
- [x] add case for nested types
- [x] add case for partition columns
- [x] UnsupportedTypeException.type -> message?
- [x] Fix failures when all columns are unsupported or specifically selecting only unsupported columns - these failure messages are expected in such cases
```java
assertQueryFails("SELECT * FROM " + tableName, "(.*)SELECT \\* not allowed from relation that has no columns");
assertQueryFails("SELECT col FROM " + tableName, "(.*)Column 'col' cannot be resolved");
```

## Additional context and related issues
https://github.com/trinodb/trino/pull/22142, fixes https://github.com/trinodb/trino/issues/22433

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by SemionPar** [comment]
While testing nested types I discovered that for `struct`, `metadata` object is misplaced:

```json
{
  "type": "struct",
  "fields": [
    {
      "name": "s",
      "type": {
        "type": "struct",
        "fields": [
          {
            "name": "field",
            "type": "double",
            "nullable": true,
            "metadata": {
              "delta.typeChanges": [
                {
                  "toType": "short",
                  "fromType": "byte",
                  "tableVersion": 2
                },
                {
                  "toType": "double",
                  "fromType": "short",
                  "tableVersion": 7
                }
              ]
            }
          }
        ]
      },
      "nullable": true,
      "metadata": {}
    },
```

VS `array` which has `metadata` where we look for it now: 

```json
    {
      "name": "a",
      "type": {
        "type": "array",
        "elementType": "double",
        "containsNull": true
      },
      "nullable": true,
      "metadata": {
        "delta.typeChanges": [
          {
            "toType": "short",
            "fromType": "byte",
            "tableVersion": 5,
            "fieldPath": "element"
          },
          {
            "toType": "double",
            "fromType": "short",
            "tableVersion": 10,
            "fieldPath": "element"
          }
        ]
      }
    }
```
I will initially implement skipping of the whole struct if one of the fields type widening is not supported.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 1]
```diff
@@ -1629,7 +1704,8 @@ public void testTypeWideningUnsupported()
         copyDirectoryContents(new File(Resources.getResource("deltalake/type_widening_unsupported").toURI()).toPath(), tableLocation);
         assertUpdate("CALL system.register_table(CURRENT_SCHEMA, '%s', '%s')".formatted(tableName, tableLocation.toUri()));
 
-        assertQueryFails("SELECT * FROM " + tableName, "Type change from 'byte' to 'unsupported' is not supported");
+        assertQueryFails("SELECT * FROM " + tableName, "(.*)SELECT \\* not allowed from relation that has no columns");
```
Comment (SemionPar) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1707]: Since now we are just skipping columns with unsupported widening, is it sane to assume that this is the expected error in such a case (query only using unsupported columns)? @ebyhr

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 2]
```diff
@@ -514,17 +512,28 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
...
-                physicalColumnType = buildType(typeManager, typeNode, true);
                 break;
             default:
                 physicalName = fieldName;
-                physicalColumnType = columnType;
         }
+        // struct cannot be a partition column
+        if (metadata.isEmpty() && isStruct(typeNode)) {
+            verifyStructTypeChanges(typeNode);
+        }
```
Comment (nineinchnick) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:525]: This should be done in `verifyTypeChanges`, or a new method. This code is slightly duplicated inside `verifyStuctTypeChanges`.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/TestDeltaLakeSchemaSupport.java [hunk 3]
```diff
@@ -111,7 +112,7 @@ public void testSinglePrimitiveFieldSchema()
 
     private void testSinglePrimitiveFieldSchema(String json, ColumnMetadata metadata)
     {
-        List<ColumnMetadata> schema = DeltaLakeSchemaSupport.getColumnMetadata(json, typeManager, ColumnMappingMode.NONE).stream()
+        List<ColumnMetadata> schema = DeltaLakeSchemaSupport.getColumnMetadata(json, typeManager, ColumnMappingMode.NONE, Collections.emptyList()).stream()
```
Comment (nineinchnick) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/TestDeltaLakeSchemaSupport.java:115]: Why not `List.of()` instead of `Collections.emptyList()`?

**Review by nineinchnick — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 4]
```diff
@@ -1547,6 +1548,42 @@ private void testReadV2Checkpoint(String resourceName)
         assertQuery("SELECT * FROM " + tableName, "VALUES (1, 2)");
     }
 
+    @Test
+    public void testTypeWideningNotSkippingUnsupportedPartitionColumns()
+            throws Exception
+    {
+        String tableName = "test_part_type_widening_" + randomNameSuffix();
+        Path tableLocation = Files.createTempFile(tableName, null);
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1556]: Please create tables under `catalogDir` so that we can remove the directory at the end of tests. Same for others. 
```suggestion
        Path tableLocation = catalogDir.resolve(tableName);
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 5]
```diff
@@ -534,7 +543,41 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
...
+    }
+
+    private static void verifyTypeChanges(JsonNode metadata)
+            throws UnsupportedTypeException
+    {
+        if (metadata.has("delta.typeChanges")) {
+            Iterator<JsonNode> typeChanges = metadata.get("delta.typeChanges").elements();
+            while (typeChanges.hasNext()) {
+                DeltaLakeSchemaSupport.verifyTypeChange(typeChanges.next());
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:574]:
```suggestion
                verifyTypeChange(typeChanges.next());
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 6]
```diff
@@ -1575,6 +1612,39 @@ public void testTypeWidening()
                 .matches("VALUES integer '127', integer '32767', integer '2147483647'");
     }
 
+    @Test
+    public void testTypeWideningNestedSkipping()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1616]: I'm trying to understand, what exactly is being skipped here?
`testTypeWideningNestedSkipping` the name of the method is not helping me.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 7]
```diff
@@ -1547,6 +1548,42 @@ private void testReadV2Checkpoint(String resourceName)
...
+
+    @Test
+    public void testTypeWideningSkippingUnsupportedColumns()
+            throws Exception
+    {
+        String tableName = "test_db_type_widening_" + randomNameSuffix();
+        Path tableLocation = catalogDir.resolve(tableName);
+        copyDirectoryContents(new File(Resources.getResource("databricks153/type_widening").toURI()).toPath(), tableLocation);
+        assertUpdate("CALL system.register_table(CURRENT_SCHEMA, '%s', '%s')".formatted(tableName, tableLocation.toUri()));
+
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1580]: Add

```
assertQuery("SELECT * FROM " + tableName + " FOR VERSION AS OF 1", "VALUES (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1.0, 1, DATE '2024-06-19')");
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 8]
```diff
@@ -471,18 +472,18 @@ public static void verifySupportedColumnMapping(ColumnMappingMode mappingMode)
...
             Iterator<JsonNode> nodes = OBJECT_MAPPER.readTree(json).get("fields").elements();
             while (nodes.hasNext()) {
                 try {
-                    columns.add(mapColumn(typeManager, nodes.next(), mappingMode));
+                    columns.add(mapColumn(typeManager, nodes.next(), mappingMode, partitionColumns));
                 }
                 catch (UnsupportedTypeException e) {
                     // Write operations are denied by unsupported 'variantType' writer feature
-                    log.debug("Skip unsupported column type: %s", e.type());
+                    log.debug("Skip unsupported column type: %s", e.getMessage());
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:486]: This change seems unrelated to the current commit.

File: plugin/trino-delta-lake/src/test/resources/databricks153/type_widening/README.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+--      * float	    -> double
+--      * decimal	-> decimal with greater precision and scale
+--      * date	    -> timestampNTZ
+--
+--  Covering all possible type changes
+--  from https://docs.databricks.com/en/delta/type-widening.html#supported-type-changes
+
+CREATE TABLE default.test_db_type_widening
+    (
+        bts BYTE, bti BYTE, btl BYTE, btd BYTE, btdb BYTE,
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/resources/databricks153/type_widening/README.md:17]: `byte_to_short` seems more readable than `bts`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 10]
```diff
@@ -471,18 +472,18 @@ public static void verifySupportedColumnMapping(ColumnMappingMode mappingMode)
...
         try {
             ImmutableList.Builder<DeltaLakeColumnMetadata> columns = ImmutableList.builder();
             Iterator<JsonNode> nodes = OBJECT_MAPPER.readTree(json).get("fields").elements();
             while (nodes.hasNext()) {
                 try {
-                    columns.add(mapColumn(typeManager, nodes.next(), mappingMode));
+                    columns.add(mapColumn(typeManager, nodes.next(), mappingMode, partitionColumns));
                 }
                 catch (UnsupportedTypeException e) {
                     // Write operations are denied by unsupported 'variantType' writer feature
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:485]: The comment is not corresponding anymore only to variant - the catch statement deals now also with unsupported type widenings.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/TestDeltaLakeSchemaSupport.java [hunk 11]
```diff
@@ -111,7 +111,7 @@ public void testSinglePrimitiveFieldSchema()
 
     private void testSinglePrimitiveFieldSchema(String json, ColumnMetadata metadata)
     {
-        List<ColumnMetadata> schema = DeltaLakeSchemaSupport.getColumnMetadata(json, typeManager, ColumnMappingMode.NONE).stream()
+        List<ColumnMetadata> schema = DeltaLakeSchemaSupport.getColumnMetadata(json, typeManager, ColumnMappingMode.NONE, List.of()).stream()
```
[the remark this answers is not in the document]
Comment (SemionPar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/TestDeltaLakeSchemaSupport.java:114]: I think we are good here, from `List.of()` javadoc: `Returns an unmodifiable list containing zero elements. See Unmodifiable Lists for details.`

File: plugin/trino-delta-lake/src/test/resources/databricks153/type_widening_partition/README.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
+Data generated using Databricks 15.3:
+
+```sql
+CREATE TABLE default.test_part_exhaustive_type_widening
```
[the remark this answers is not in the document]
Comment (SemionPar) [reply plugin/trino-delta-lake/src/test/resources/databricks153/type_widening_partition/README.md:4]: Drift is unintentional, it should be named `test_type_widening_partition` according to the convention I see in other directories - changing this. 
The name in the readme is not taken into account by the test, though - the directory name matters.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java (hunk 6 above)
Comment (SemionPar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1616]: Does it sound better as`testTypeWideningSkippingUnsupportedNestedTypes`? This test asserts that columns of nested types where unsupported widening happens on one of the nested fields are skipped.

File: plugin/trino-delta-lake/src/test/resources/databricks153/type_widening/README.md (hunk 9 above)
Comment (SemionPar) [reply plugin/trino-delta-lake/src/test/resources/databricks153/type_widening/README.md:17]: Agree, what was I thinking? I will rename these in a separate push.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 8 above)
Comment (SemionPar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:486]: Hmm, yes, I can revert and do exception class change in a separate commit.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java [hunk 13]
```diff
@@ -167,7 +168,14 @@ public static Object deserializePartitionValue(DeltaLakeColumnHandle column, Opt
...
     {
-        LocalDateTime localDateTime = LocalDateTime.parse(timestamp, PARTITION_TIMESTAMP_FORMATTER);
+        LocalDateTime localDateTime;
+        try {
+            localDateTime = LocalDateTime.parse(timestamp, PARTITION_TIMESTAMP_FORMATTER);
+        }
+        catch (DateTimeParseException ignored) {
+            // type widening date->timestamp has occurred and this value is of date format
+            localDateTime = LocalDate.parse(timestamp).atTime(LocalTime.MIN);
+        }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java:178]: Ideally, we should avoid this exception-driven logic. Could you leave a TODO comment?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 14]
```diff
@@ -106,7 +106,10 @@ public class TestDeltaLakeBasic
             new ResourceTable("uniform_iceberg_v2", "databricks143/uniform_iceberg_v2"),
             new ResourceTable("unsupported_writer_feature", "deltalake/unsupported_writer_feature"),
             new ResourceTable("unsupported_writer_version", "deltalake/unsupported_writer_version"),
-            new ResourceTable("variant", "databricks153/variant"));
+            new ResourceTable("variant", "databricks153/variant"),
+            new ResourceTable("test_db_type_widening", "databricks153/type_widening"),
+            new ResourceTable("test_db_partition_type_widening", "databricks153/type_widening_partition"),
+            new ResourceTable("test_db_type_widening_nested", "databricks153/type_widening_nested"));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:112]: Remove `test_db_` prefix for consistency with other table names. No need to recreate tables. Just updating the name here and README.md should be fine.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 15]
```diff
@@ -1547,6 +1550,31 @@ private void testReadV2Checkpoint(String resourceName)
...
+        assertThat(query("SELECT * FROM test_db_partition_type_widening FOR VERSION AS OF 1"))
+                .matches("VALUES (tinyint '1', tinyint '1', tinyint '1', tinyint '1', tinyint '1', tinyint '1', smallint '1', smallint '1', smallint '1', smallint '1', integer '1', integer '1', integer '1', bigint '1', real '1', CAST('1' AS decimal(10,0)), DATE '2024-06-19')");
+        assertThat(query("SELECT * FROM test_db_partition_type_widening FOR VERSION AS OF 18"))
+                .matches("VALUES (tinyint '1', smallint '1', integer '1', bigint '1', CAST('1' AS decimal(10,0)), double '1', integer '1', bigint '1', CAST('1' AS decimal(10,0)), double '1', bigint '1', CAST('1' AS decimal(10,0)), double '1', CAST('1' AS decimal(20,0)), double '1', CAST('1' AS decimal(12,2)), CAST('2024-06-19' AS timestamp(6))), " +
+                        "(tinyint '2', smallint '2', integer '2', bigint '2', CAST('2' AS decimal(10,0)), double '2', integer '2', bigint '2', CAST('2' AS decimal(10,0)), double '2', bigint '2', CAST('2' AS decimal(10,0)), double '2', CAST('2' AS decimal(20,0)), double '2', CAST('2.22' AS decimal(12,2)), TIMESTAMP '2021-07-01 08:43:28.123456')");
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1565]: These lines look too long. Could you wrap per each value?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 16]
```diff
@@ -1547,6 +1550,31 @@ private void testReadV2Checkpoint(String resourceName)
         assertQuery("SELECT * FROM " + tableName, "VALUES (1, 2)");
     }
 
+    @Test
+    public void testTypeWideningNotSkippingUnsupportedPartitionColumns()
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:1554]: Could you add javadoc with link to resource path so that we can confirm the transaction logs quickly? Same for other tests. 

Also, I would move testTypeWideningNotSkippingUnsupportedPartitionColumns and testTypeWideningSkippingUnsupportedColumns under testTypeWidening.

File: plugin/trino-delta-lake/src/test/resources/databricks153/type_widening_partition/README.md [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+INSERT INTO default.test_db_partition_type_widening VALUES
+    (
+        2,
+        2, 2, 2, 2, 2,
+        2, 2, 2, 2,
+        2, 2, 2,
+        2,
+        2.0,
+        2.22,
+        TIMESTAMP_NTZ '2021-7-1T8:43:28.123456'
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/resources/databricks153/type_widening_partition/README.md:58]: Ideally, we should insert values that old types can't represent. For instance, the new value for `Byte_to_Short` should be larger/smaller than the byte range. 

Additional values aren't important in `databricks153/type_widening` & `databricks153/type_widening_nested` because of unsupported type changes, but it's little important in `databricks153/type_widening_partition` because we want to arrow reading those values.

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 18]
```diff
@@ -514,17 +513,21 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
...
-                physicalColumnType = columnType;
         }
+        verifyTypeChanges(metadata, typeNode, partitionColumns.contains(physicalName));
+        physicalColumnType = switch (mappingMode) {
+            case ID, NAME -> {
+                verify(!isNullOrEmpty(physicalName), "physicalName is null or empty");
+                yield buildType(typeManager, typeNode, true);
+            }
+            default -> columnType;
+        };
```
[the remark this answers is not in the document]
Comment (SemionPar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:530]: `physicalName` is needed to check if column is a partition column for `verifyTypeChanges`, so I moved resolving of `physicalName` before `verifyTypeChanges`. Seeing that `buildType` part of the `physicalName` resolving `switch` is 'busy', I extracted it and placed it after `verifyTypeChanges`, so we don't perform it for columns we are skipping - if you think this is a misguided optimization, please 👎  and I will move it back into a single switch.

File: plugin/trino-delta-lake/src/test/resources/databricks153/type_widening_partition/README.md (hunk 17 above)
Comment (SemionPar) [reply plugin/trino-delta-lake/src/test/resources/databricks153/type_widening_partition/README.md]: Thank you, changed inserted values for `databricks153/type_widening_partition` to extend beyond original type.

**Comment by SemionPar** [comment]
Squashed commits.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java (hunk 13 above)
Comment (SemionPar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogParser.java]: Yes, something like appending another pattern?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 19]
```diff
@@ -503,28 +505,28 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
...
             case NAME:
                 physicalName = metadata.get("delta.columnMapping.physicalName").asText();
-                verify(!isNullOrEmpty(physicalName), "physicalName is null or empty");
-                physicalColumnType = buildType(typeManager, typeNode, true);
                 break;
             default:
                 physicalName = fieldName;
-                physicalColumnType = columnType;
         }
+        verifyTypeChanges(metadata, typeNode, partitionColumns.contains(physicalName));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:522]: > partitionColumns.contains(physicalName)

This condition seems wrong. `partitionColumns` holds logical column names. You can check with the following steps:
```sh
testing/bin/ptl env up --environment EnvSinglenodeDeltaLakeOss

docker exec -it ptl-spark spark-sql

CREATE TABLE test_part(a int, b int) USING delta location 's3://test-bucket/test_part' PARTITIONED BY (b) TBLPROPERTIES ('delta.columnMapping.mode' = 'name');

Open http://localhost:9080/ (minio-access-key & minio-secret-key)
```

Probably, we could change the argument to `fieldName`, move the verifyTypeChanges call before `switch (mappingMode)` and revert physicalColumnType change:
```java
        verifyTypeChanges(metadata, typeNode, partitionColumns.contains(fieldName));
        switch (mappingMode) {
...
```
